mubench.pipeline/data/build_command.py

```diff
@@ -0,0 +1,190 @@
+import logging
+import shutil
+import os
+import shlex
+from glob import glob
+from os import makedirs
+from os.path import join, exists, dirname, splitext, relpath
+from typing import List, Set
+
+from utils.io import remove_tree, copy_tree
+from utils.shell import Shell, CommandFailedError
+
+
+class BuildCommand:
+    def __init__(self, name: str, args: List[str]):
+        self.logger = logging.getLogger("buildcommand")
+        self._name = name
+        self.args = args
+
+    @staticmethod
+    def get(requested_command: str) -> 'BuildCommand':
+        requested_command = shlex.split(requested_command)
+        requested_name = requested_command[0]
+        args = requested_command[1:]
+
+        commands = BuildCommand._get_implementations()
+        matching_commands = [c for c in commands if
+                             c.name() == requested_name]
+        if len(matching_commands) == 1:
+            matching_command = matching_commands[0]
+            return matching_command(requested_name, args)
+        if not matching_commands:
+            return BuildCommand(requested_name, args)
+
+        raise ValueError("Multiple matching commands found for {}"
+                         .format(requested_command))
+
+    @staticmethod
+    def _get_implementations():
+        return BuildCommand.__subclasses__()
+
+    def execute(self, project_dir: str, dep_dir: str, compile_base_path: str) -> None:
+        command = self._get_command(self.args)
+
+        try:
+            output = Shell.exec(command, cwd=project_dir, logger=self.logger)
+        except CommandFailedError as e:
+            raise CommandFailedError(e.command, '\n' + self._get_errors(e.output))
+
+        self._copy_dependencies(output, project_dir, dep_dir, compile_base_path)
+
+    @staticmethod
+    def name() -> str:
+        raise NotImplementedError
+
+    def _get_command(self, args: List[str]) -> str:
+        return ' '.join([self._name] + self._extend_args(args))
+
+    def _extend_args(self, args: List[str]) -> List[str]:
+        return args
+
+    def _get_errors(self, output: str) -> str:
+        return output
+
+    def _copy_dependencies(self, exec_output: str, project_dir: str, dep_dir: str, compile_base_path: str) -> None:
```

Comment:
All three implementations of this method end with the same two lines: retrieve dependency list and copy dependencies. It would be more consistent and a little less repetition if this would be `_get_dependencies`, which returns the dependency list and the copy would happen in the caller. This leads us to the problem that there are commands without dependencies, where the copy might fail (it wouldn't, currently, but I don't like to depend on this). I think the clean solution would be to have a subclass of `BuildCommand`, say, `BuildToolCommand`, which only adds the dependency handling and is the superclass of the three specialized commands we currently have... might also be overkill and we can just return an empty list by default and not invoke the copy if the list is empty. What do you think?